QAchat/python/QAchat.ipynb

```diff
@@ -0,0 +1,428 @@
+    "            output=testCase['output'],\n",
+    "        )\n",
+    "\n",
+    "# List of facts about a fake user. These will be given as context to the QAChatbot.\n",
+    "userContextDataset = []\n",
+    "with open('userContextDataset.jsonl') as f:\n",
+    "    for line in f:\n",
+    "        userContextDataset.append(json.loads(line))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Store the user context dataset\n",
+    "Next, we will embed the personal dataset of information into a vector DB. We will use [Chroma](https://www.trychroma.com/) to embed, store, and retrieve the relevant user context."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "chroma_client = chromadb.Client()\n",
+    "collection = chroma_client.create_collection(name=\"user-context\")\n",
+    "\n",
+    "for i,c in enumerate(userContextDataset):\n",
+    "    fact = c[\"detail\"]\n",
+    "    collection.add(\n",
+    "        documents = [fact],\n",
+    "        ids = [str(i)]\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Write an evaluation function\n",
+    "\n",
+    "Now, we will define a general evaluation function to test different prompts and pipelines using BrainTrust. This makes it easy to iterate and improve our AI app."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "# This is an evaluation script that tests a prompt and pipeline combination and logs the results to BrainTrust.\n",
+    "def runEvaluation(generationFn, experimentName):\n",
+    "    # Initialize a BrainTrust experiment\n",
+    "    experiment = braintrust.init(project=PROJECTNAME, api_key=BT_API_KEY, experiment=experimentName, dataset=dataset)\n",
+    " \n",
+    "    for testCase in dataset:\n",
+    "        input_data = str(testCase[\"input\"])\n",
+    "        expected = str(testCase[\"output\"])\n",
+    "    \n",
+    "        output, prompt =  generationFn(input_data) # Generate an answer using the pipeline given\n",
+    " \n",
+    "        # Score factuality using BrainTrust's factuality scorer\n",
+    "        # Learn more here: https://github.com/braintrustdata/autoevals/blob/main/templates/factuality.yaml\n",
+    "        factuality = Factuality()\n",
+    "        factualityScore = factuality(output, expected, input=input_data)\n",
+    "\n",
+    "        # Score the output using BrainTrust's Levenshtein scorer\n",
+    "        levenEvaluator = LevenshteinScorer()\n",
+    "        levenScore = levenEvaluator(output, expected, input=input_data)\n",
+    "        \n",
+    "        # Define two simple custom scorers to ensure the LLM is saying the right words\n",
+    "        def BadWordScorer(output):\n",
```

Comment:
Why are these function names capitalized? Also, place them outside the definition of this function